File: chatbot/src/adapters/site_b/mappers.py
from typing import Any, Dict, List
from ..schema import (
    User,
    ProductSearchResult,
    ProductSummary,
    Money,
    GetOrderStatusResult,
    OrderSummary,
    OrderItem,
    GetDeliveryTrackingResult,
    DeliveryTracking,
    SubmitOrderActionResult,
)


def to_order_status_token(raw: Any) -> str:
    val = str(raw or "unknown").lower()
    if "결제" in val and "대기" in val:
        return "pending"
    if "결제" in val and "완료" in val:
        return "paid"
    if "배송" in val and "준비" in val:
        return "preparing"
    if "배송" in val and "중" in val:
        return "shipped"
    if "배송" in val and "완료" in val:
        return "delivered"
    if "취소" in val:
        return "cancelled"
    if "환불" in val:
        return "refunded"
    return val


def to_delivery_status_token(raw: Any) -> str:
    val = str(raw or "unknown").lower()
    if "준비" in val:
        return "ready"
    if "배송" in val and "중" in val:
        return "in_transit"
    if "완료" in val:
        return "delivered"
    return val


def map_site_b_user(raw: Any, site_id: str) -> User:
    orders = raw.get("orders", [])
    user_id = str(orders[0].get("user_id", "")) if orders else ""
    return User(id=user_id, siteId=site_id, email=None, name=None)


def map_site_b_product_search(raw: Any, site_id: str) -> ProductSearchResult:
    products = raw.get("products", [])
    items: List[ProductSummary] = []

    for item in products:
        price_obj = None
        if item.get("price") is not None:
            price_obj = Money(amount=float(item.get("price")), currency="KRW")

        items.append(
            ProductSummary(
                id=str(item.get("product_id")),
                siteId=site_id,
                title=item.get("name", ""),
                shortDescription=item.get("description"),
                price=price_obj,
                inStock=(item.get("stock", 0) > 0),
                imageUrl=item.get("image_url"),
                categoryIds=[str(item.get("category"))]
                if item.get("category")
                else None,
                brand=item.get("brand"),
            )
        )

    return ProductSearchResult(items=items, total=len(items))


def map_site_b_order(raw: Any, deps: Dict[str, Any]) -> GetOrderStatusResult:
    site_id = deps["site_id"]
    current_user_id = deps["current_user_id"]
    target_order_id = deps["target_order_id"]
    norm_order_status = deps["normalize_order_status"]

    orders = raw.get("orders", [])
    found_order = next(
        (o for o in orders if str(o.get("order_id")) == target_order_id),
        orders[0] if orders else {},
    )
    order_id = found_order.get("order_id", "")

    items = []
    for item in found_order.get("items", []):
        unit_price = (
            Money(amount=float(item.get("price")), currency="KRW")
            if item.get("price") is not None
            else None
        )
        items.append(
            OrderItem(
                productId=str(item.get("product_id")),
                productTitle=item.get("product_name", ""),
                quantity=int(item.get("quantity", 0)),
                unitPrice=unit_price,
                imageUrl=item.get("image_url"),
            )
        )

    total_price = (
        Money(amount=float(found_order.get("total_price")), currency="KRW")
        if found_order.get("total_price") is not None
        else None
    )

    order = OrderSummary(
        orderId=str(order_id),
        siteId=site_id,
        userId=current_user_id,
        status=norm_order_status(
            to_order_status_token(found_order.get("status", "unknown"))
        ),
        items=items,
        totalPrice=total_price,
        orderedAt=found_order.get("created_at"),
    )
    return GetOrderStatusResult(order=order)


def map_site_b_delivery(raw: Any, deps: Dict[str, Any]) -> GetDeliveryTrackingResult:
    target_order_id = deps["target_order_id"]
    orders = raw.get("orders", [])
    found_order = next(
        (o for o in orders if str(o.get("order_id")) == target_order_id),
        orders[0] if orders else {},
    )
    order_id = str(found_order.get("order_id", ""))

    tracking = DeliveryTracking(
        orderId=order_id,
        deliveryStatus=deps["normalize_delivery_status"](
            to_delivery_status_token(found_order.get("status", "unknown"))
        ),
        lastUpdatedAt=found_order.get("created_at"),
    )
    return GetDeliveryTrackingResult(tracking=tracking)


def map_site_b_order_action(raw: Any) -> SubmitOrderActionResult:
    return SubmitOrderActionResult(
        success=False,
        status="not_allowed",
        message="bilyeo 사이트는 주문 액션 API를 제공하지 않습니다.",
    )

File: SaaS/src/adapters/site-b/mappers.ts
import {
  GetDeliveryTrackingResult,
  GetOrderStatusResult,
  KnowledgeSearchResult,
  ProductSearchResult,
  SubmitOrderActionResult,
  User
} from "../../domain/ecommerce";

type SiteBMapperDeps = {
  siteId: string;
  currentUserId: string;
  targetOrderId: string;
  normalizeOrderStatus: (raw: string) => GetOrderStatusResult["order"]["status"];
  normalizeDeliveryStatus: (
    raw: string
  ) => GetDeliveryTrackingResult["tracking"]["deliveryStatus"];
};

const toOrderStatusToken = (raw: unknown): string => {
  const value = String(raw ?? "unknown").toLowerCase();
  if (value.includes("결제") && value.includes("대기")) return "pending";
  if (value.includes("결제") && value.includes("완료")) return "paid";
  if (value.includes("배송") && value.includes("준비")) return "preparing";
  if (value.includes("배송") && value.includes("중")) return "shipped";
  if (value.includes("배송") && value.includes("완료")) return "delivered";
  if (value.includes("취소")) return "cancelled";
  if (value.includes("환불")) return "refunded";
  return value;
};

const toDeliveryStatusToken = (raw: unknown): string => {
  const value = String(raw ?? "unknown").toLowerCase();
  if (value.includes("준비")) return "ready";
  if (value.includes("배송") && value.includes("중")) return "in_transit";
  if (value.includes("완료")) return "delivered";
  return value;
};

export function mapSiteBUser(raw: any, siteId: string): User {
  const orders = raw.orders ?? [];

  return {
    id: String(orders[0]?.user_id ?? ""),
    siteId,
    email: undefined,
    name: undefined
  };
}

export function mapSiteBProductSearch(raw: any, siteId: string): ProductSearchResult {
  const products = raw.products ?? [];

  return {
    items: products.map((item: any) => ({
      id: String(item.product_id),
      siteId,
      title: item.name,
      shortDescription: item.description,
      price: item.price !== undefined
          ? {
              amount: Number(item.price),
              currency: "KRW"
            }
          : undefined,
      inStock: (item.stock ?? 0) > 0,
      imageUrl: item.image_url,
      categoryIds: item.category ? [String(item.category)] : undefined,
      brand: item.brand
    })),
    total: products.length
  };
}

export function mapSiteBKnowledge(raw: any, siteId: string): KnowledgeSearchResult {
  void raw;
  void siteId;
  return {
    documents: []
  };
}

export function mapSiteBOrder(raw: any, deps: SiteBMapperDeps): GetOrderStatusResult {
  const orders = raw.orders ?? [];
  const found =
    orders.find((o: any) => String(o.order_id) === String(deps.targetOrderId)) ?? orders[0];
  const order = found ?? {};

  return {
    order: {
      orderId: String(order.order_id ?? ""),
      siteId: deps.siteId,
      userId: deps.currentUserId,
      status: deps.normalizeOrderStatus(toOrderStatusToken(order.status ?? "unknown")),
      items: (order.items ?? []).map((item: any) => ({
        productId: String(item.product_id),
        productTitle: item.product_name,
        quantity: Number(item.quantity),
        unitPrice: item.price !== undefined
          ? {
              amount: Number(item.price),
              currency: "KRW"
            }
          : undefined,
        imageUrl: item.image_url
      })),
      totalPrice: order.total_price !== undefined
        ? {
            amount: Number(order.total_price),
            currency: "KRW"
          }
        : undefined,
      orderedAt: order.created_at
    }
  };
}

export function mapSiteBDelivery(
  raw: any,
  deps: SiteBMapperDeps
): GetDeliveryTrackingResult {
  const orders = raw.orders ?? [];
  const order =
    orders.find((o: any) => String(o.order_id) === String(deps.targetOrderId)) ?? orders[0] ?? {};

  return {
    tracking: {
      orderId: String(order.order_id ?? ""),
      deliveryStatus: deps.normalizeDeliveryStatus(toDeliveryStatusToken(order.status ?? "unknown")),
      lastUpdatedAt: order.created_at
    }
  };
}

export function mapSiteBOrderAction(raw: any): SubmitOrderActionResult {
  void raw;
  return {
    success: false,
    status: "not_allowed",
    message: "bilyeo 사이트는 주문 액션 API를 제공하지 않습니다."
  };
}
